src/internal/middleware.js

```diff
@@ -37,20 +37,24 @@ export default function sagaMiddlewareFactory(options = {}) {
   function sagaMiddleware({getState, dispatch}) {
     runSagaDynamically = runSaga
     const sagaEmitter = emitter()
+    const sagaDispatch = !sagaMonitor ? dispatch : wrapSagaDispatch(dispatch)
```

Comment:
wondering if distinguishing permanently regular `dispatch`es and `put`s could be leveraged to handle `asap` queue better in order to fix cases like - https://github.com/yelouafi/redux-saga/issues/598